cpp/lexer/lexer.cc

```diff
@@ -0,0 +1,92 @@
+#include "lexer.hh"
+
+
+#include <exception>
+
+
+lexer::lexer(const std::string input)
+  : input_{input},
```

Comment:
Wouldn't it make more sense to pass the `input` argument either by `const&` reference or non-`const` value then `std::move()`-ing it into `input_`? It seems like passing by `const` value will guarantee to create an extra copy of the input string, unless I'm mistaken.

I haven't read the book this is based on (yet), but assuming the input string outlives the lexer, it could also use `std::string_view` to avoid any copies at all.

Anyway, just happened here after watching a Primeagen video about this project and was glad to see someone representing C++, nice work! :+1: